src/libstore/file-hash.hh

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <variant>
+#include "hash.hh"
+
+namespace nix {
+
+enum struct FileIngestionMethod : uint8_t {
+    Flat = false,
+    Recursive = true
+};
+
+struct TextHash {
+    Hash hash;
+    TextHash(const TextHash &) = default;
+    TextHash(TextHash &&) = default;
+    TextHash & operator = (const TextHash &) = default;
+};
+
+/// Pair of a hash, and how the file system was ingested
+struct FileSystemHash {
+    FileIngestionMethod method;
```

Comment:
this might need to be optional

Reply:
I don't think so?

Reply:
Mainly because of the above comment. Not all uses of FileSystemHash will have a real "FileIngestionMethod" defined. Consider derivations.

Reply:
Oh, we should rename the data type then. This is strictly for content-addressed things.